### PR TITLE
Fix/multiclass recall macro avg ignore index 

### LINKED ISSUE #2441: Error in computing `MulticlassRecall` macro average when `ignore_index` is specified
## 🐛 Bug

In `MulticlassRecall` the `ignore_index` argument leads to incorrect behaviour when computing the macro average. It seems as if the metric still divides by the total number of classes rather than ignoring the `ignore_index` class (while treating the `ignore_index` class as having a score of `0`).

### To Reproduce

```python
import torch
from torchmetrics.classification.precision_recall import MulticlassRecall

metric = MulticlassRecall(num_classes=2, ignore_index=0, average="macro")
y_true = torch.tensor([0, 0, 1, 1])
y_pred = torch.tensor([
    [0.9, 0.1],  # Correct
    [0.9, 0.1],  # Correct
    [0.9, 0.1],  # Incorrect
    [0.1, 0.9],  # Correct
])
metric.update(y_pred, y_true)
print(metric.compute())  # Prints 0.25, but I would expect the result to be 0.5
```

### Expected behavior

I would expect the recall score to be `0.5`.

In the toy example above I have two classes (0 and 1) and two instances of each class. I want to ignore class 0 altogether (hence `ignore_index=0`). The first prediction for class 1 (the third prediction overall) is incorrect (it predicts class 0 with confidence 0.9 instead), but the other prediction is correct. Therefore, I would expect the recall score to be `0.5`. 

However, the actual result is `0.25`. It looks as if this metric is still dividing by the total number of classes when computing the macro average instead of ignoring the `ignore_index` class (while treating the `ignore_index` class as having a score of `0`).

(Notice that the predictions for class 0 are in fact correct. If I leave out the `ignore_index=0` altogether, I get an expected result of `0.75`.)

### Environment

- TorchMetrics version: 1.3.1 (installed via pip)
- Python version: 3.10
- PyTorch version: 2.2.1
- OS: Ubuntu 22.04

### PR DESCRIPTION
**Pull request by rittik9**
## What does this PR do?

Fixes  #2441

<details>
  <summary>Details</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

## Did you have fun?

Yes

## Issue:
- The root of the problem seems to be that the `ignore_index` information is not being properly propagated to the final averaging step i.e. the `_adjust_weights_safe_divide` function doesn't know that which class should be ignored.
## Solution:
- To address this issue, I updated the code to ensure that the `ignore_index` information is preserved throughout the entire process, making sure it is correctly passed through all intermediate steps up to the final averaging stage i.e. `_adjust_weights_safe_divide` function .
- Updated the `_adjust_weights_safe_divide` function to accept an additional `ignore_index` parameter, which is passed through the `_precision_recall_reduce` function, called in the `compute` method of the `MulticlassRecall` class. This change adjusts the weights in the  `_adjust_weights_safe_divide` function, setting the weight of the ignored class to 0.

### CODE REVIEW

**Comment by rittik9** [comment]
> looks good, can we add also test for this case...

Sure

**Review by Borda — DISMISSED** [review]

**Comment by rittik9** [comment]
@Borda  What do I have to modify?

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
@rittik9 mind checking the changed docstest values and whether it is correct?

**Comment by rittik9** [comment]
> @rittik9 mind checking the changed docstest values and whether it is correct?

@Borda I checked the tests , I believe they are correct

**Comment by JeroenMandersloot** [comment]
Any update on when this PR could be merged? It would really help if we could update from the 0.9.3 version once this fix is merged.

**Comment by Borda** [comment]
> Any update on when this PR could be merged? It would really help if we could update from the 0.9.3 version once this fix is merged.

the tests/doctests need to be fixed, are you interested in submitting a suggestion on what else needs to be fixed/chnaged?

File: src/torchmetrics/functional/classification/accuracy.py [hunk 1]
```diff
@@ -85,7 +85,9 @@ def _accuracy_reduce(
         return _safe_divide(tp, tp + fn)
 
     score = _safe_divide(tp + tn, tp + tn + fp + fn) if multilabel else _safe_divide(tp, tp + fn)
-    return _adjust_weights_safe_divide(score, average, multilabel, tp, fp, fn, top_k)
```
Comment (Borda) [inline src/torchmetrics/functional/classification/accuracy.py:88]: seems we had this bug with passing `top_k` as index :(

File: src/torchmetrics/utilities/compute.py [hunk 2]
```diff
@@ -61,7 +61,14 @@ def _safe_divide(num: Tensor, denom: Tensor, zero_division: float = 0.0) -> Tens
...
 def _adjust_weights_safe_divide(
-    score: Tensor, average: Optional[str], multilabel: bool, tp: Tensor, fp: Tensor, fn: Tensor, top_k: int = 1
+    score: Tensor,
+    average: Optional[str],
+    multilabel: bool,
+    tp: Tensor,
+    fp: Tensor,
+    fn: Tensor,
+    ignore_index: Optional[int] = None,
+    top_k: int = 1,
```
Comment (Borda) [inline src/torchmetrics/utilities/compute.py:71]:
```suggestion
    top_k: int = 1,
    ignore_index: Optional[int] = None,
```

File: src/torchmetrics/functional/classification/precision_recall.py [hunk 3]
```diff
@@ -56,7 +57,9 @@ def _precision_recall_reduce(
         return _safe_divide(tp, tp + different_stat, zero_division)
 
     score = _safe_divide(tp, tp + different_stat, zero_division)
-    return _adjust_weights_safe_divide(score, average, multilabel, tp, fp, fn, top_k=top_k)
+    return _adjust_weights_safe_divide(
+        score=score, average=average, multilabel=multilabel, tp=tp, fp=fp, fn=fn, ignore_index=ignore_index, top_k=top_k
```
Comment (Borda) [inline src/torchmetrics/functional/classification/precision_recall.py:61]: @rittik9 mind checking what is wrong here, seems that the computed metrics do not match the reference...

File: tests/unittests/classification/test_precision_recall.py [hunk 4]
```diff
@@ -661,6 +661,37 @@ def test_corner_case():
     assert res == 1.0
 
 
+def test_multiclass_recall_ignore_index():
```
[the remark this answers is not in the document]
Comment (Borda) [reply tests/unittests/classification/test_precision_recall.py:664]: looking to the code and the ignore index is already applied in `_multilabel_stat_scores_format` which reduces the preds/target size the same way as the reference metric so calling it with null weights in fact ignores additional index

[the remark this answers is not in the document]
Comment (Borda) [reply tests/unittests/classification/test_precision_recall.py:664]: well getting to the root, so the problem is when you try to use the class metric and there is no update called...

[the remark this answers is not in the document]
Comment (rittik9) [reply tests/unittests/classification/test_precision_recall.py]: The problem is we are using sklearn's ```recall_score``` as a reference for our unittests. But sklearn's ```recall_score``` does not support ```ignore_index``` argument. So even if in ```_reference_sklearn_precision_recall_multiclass()``` function we are using ```remove_ignore_index``` function for removing those predictions whose real values are ```ignore_index``` class before passing it to ```recall_score``` function, it does not matter. Because whenever ```average='macro'``` sklearn's ```recall_score``` will always return mean cosidering the total no. of classes. That is the reason why unittests passed in the first place.I think we need to figure out how we want ```MulticlassRecall()``` to behave when ```average = 'macro'``` and ```ignore_index``` is specified like this issue says (in that case we need to modify the unittests) or like it is implemented according to sklearn. I've prepared a [notebook ](https://colab.research.google.com/drive/1YyEZav4b3BY8w48Is8EceSNFoAnYgzXZ?usp=sharing)for explanation. @Borda pls take a look.

[the remark this answers is not in the document]
Comment (rittik9) [reply tests/unittests/classification/test_precision_recall.py]: The problem is we are using sklearn's ```recall_score``` as a reference for our unittests. So even if in ```_reference_sklearn_precision_recall_multiclass()``` function we are using ```remove_ignore_index``` function for removing those predictions whose real values are ```ignore_index``` class before passing it to ```recall_score``` function, it does not matter. Because whenever ```average='macro'``` sklearn's ```recall_score``` will always return mean cosidering the total no. of classes (as we are passing all the classes in ```recall_score()``` function's ```labels``` argument). That is the reason why unittests passed in the first place. I think we need to fix the unittests to take care of ignore_index using sklearn's ```recall_score()``` function's ```labels``` argument. I've prepared a [notebook](https://colab.research.google.com/drive/1YyEZav4b3BY8w48Is8EceSNFoAnYgzXZ?usp=sharing) for explanation. @Borda pls take a look.

**Comment by DimitrisMantas** [comment]
Just to chime in, I think this issue is present in pretty much all metrics that make use of `_adjust_weights_safe_divide`.

I see this PR fixes, some of them, but others, such as `JaccardIndex` are left as is.

**Comment by rittik9** [comment]
I think we gotta fix the unit tests first
